### PR TITLE
CI: test pyarrow 2.0 and nightly

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Alternative to https://github.com/dask/dask/pull/6936 (cc @jrbourbeau): since there is an upstream-dev build, we can actually install pyarrow nightly in there (instead of only doing it in a test PR as #6936  does).
Additionally, I also removed the ``!=2.0.0`` pin (which will require some skips in the tests (TODO), but I think it's still worth the run the rest of the tests with the latest released pyarrow version).

### CODE REVIEW

File: .github/workflows/ci-upstream.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ on:
 jobs:
   build:
     runs-on: "ubuntu-latest"
-    if: "contains(github.event.head_commit.message, 'test-upstream') || (github.repository == 'dask/dask' && github.event_name != 'pull_request')"
+    # if: "contains(github.event.head_commit.message, 'test-upstream') || (github.repository == 'dask/dask' && github.event_name != 'pull_request')"
```
Comment (jorisvandenbossche) [inline .github/workflows/ci-upstream.yml:12]: TODO: undo this change at the end

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@ jobs:
   test:
     runs-on: ${{ matrix.os }}
     strategy:
-      fail-fast: true
+      fail-fast: false
```
Comment (jorisvandenbossche) [inline .github/workflows/ci.yml:9]: TODO: undo this change at the end

File: continuous_integration/environment-3.6.yaml [hunk 3]
```diff
@@ -22,7 +22,7 @@ dependencies:
   - zarr
   - tiledb-py
   - sqlalchemy
-  - pyarrow>=0.14.0,!=2.0.0
+  - pyarrow=0.14.*
```
Comment (jorisvandenbossche) [inline continuous_integration/environment-3.6.yaml:25]: I pinned this to 0.14, because that is what in practice for all platforms gets used (and by putting it in the env file, it's a bit easier to see this)

File: continuous_integration/environment-3.7.yaml [hunk 4]
```diff
@@ -19,7 +19,7 @@ dependencies:
   - tiledb-py
   - fsspec>=0.6.0
   - sqlalchemy
-  - pyarrow>=0.14.0,!=2.0.0
+  - pyarrow=1.0.*
```
Comment (jorisvandenbossche) [inline continuous_integration/environment-3.7.yaml:22]: same here, this is the version that got used in practice

File: continuous_integration/environment-3.8-dev.yaml [hunk 5]
```diff
@@ -7,7 +7,7 @@ dependencies:
   - python=3.8.*
   # For the dependencies. These will be re-installed later.
   - numpy
-  - pandas
+  - pandas=1.0.*
```
Comment (jorisvandenbossche) [inline continuous_integration/environment-3.8-dev.yaml:10]: pinned here to older pandas, since otherwise I get the failures that are being handled by https://github.com/dask/dask/pull/6996

File: continuous_integration/environment-3.8.yaml [hunk 6]
```diff
@@ -19,7 +19,7 @@ dependencies:
   - tiledb-py
   - fsspec>=0.6.0
   - sqlalchemy
-  - pyarrow>=0.14.0,!=2.0.0
+  - pyarrow>=0.14.0
```
Comment (jorisvandenbossche) [inline continuous_integration/environment-3.8.yaml:22]: Here just leave the "larger than" constraint, to get the last released version of pyarrow, which at the moment is 2.0.0 (after removing the `!=2.0.0` pin)

File: setup.cfg [hunk 7]
```diff
@@ -36,10 +36,10 @@ addopts = -rsx --durations=10
...
     # From Cython-1753
     ignore:can't resolve:ImportWarning
-    error:::dask[.*]
-    error:::pandas[.*]
-    error:::numpy[.*]
-    error:::distributed[.*]]
+    # error:::dask[.*]
+    # error:::pandas[.*]
+    # error:::numpy[.*]
+    # error:::distributed[.*]]
```
Comment (jorisvandenbossche) [inline setup.cfg:42]: TODO: undo this change in the end

**Comment by jorisvandenbossche** [comment]
cc @jsignell apart from the graphviz edit, and some edits I need to do undo at the end (if all is OK otherwise), I think this should be ready

In the upstream build, remaining failure is a `sparse` related one

**Comment by jsignell** [comment]
Great!

> In the upstream build, remaining failure is a sparse related one

Yeah I think I'd done some fixes around that one in the past. Let me open a fresh PR for that.

**Comment by jorisvandenbossche** [comment]
OK, everything seems to be passing here (except the 1 sparse failure), will revert the temporary edits

File: continuous_integration/scripts/install.sh [hunk 8]
```diff
@@ -7,11 +7,14 @@ set -xe
...
-    conda uninstall --force numpy pandas
-    python -m pip install --no-deps --pre \
-        -i https://pypi.anaconda.org/scipy-wheels-nightly/simple \
-        numpy
-    python -m pip install --pre pandas==1.1.0rc0
+    # conda uninstall --force numpy pandas
+    # python -m pip install --no-deps --pre \
+    #     -i https://pypi.anaconda.org/scipy-wheels-nightly/simple \
+    #     numpy
+    # python -m pip install --pre pandas==1.1.0rc0
```
Comment (jrbourbeau) [inline continuous_integration/scripts/install.sh:14]: Apologies for not catching this earlier, but I think we want to keep this for testing nightly versions of numpy and pandas

Though I just noticed we're pinning `pandas==1.1.0rc0` which might be an oversight. Instead I would have expected us to install both numpy and pandas from https://pypi.anaconda.org/scipy-wheels-nightly/simple

Comment (jorisvandenbossche) [reply continuous_integration/scripts/install.sh:14]: Yes, we certainly want to keep that, but the thing is that they were currently failing a lot, and thus not very useful (hard to see what errors still need to be fixed for a given dependency). The pandas ones are being fixed in https://github.com/dask/dask/pull/6996 (and the pandas nightly install is added back there), and the numpy ones in https://github.com/dask/dask/pull/6896 I think

Comment (jrbourbeau) [reply continuous_integration/scripts/install.sh:14]: Great! Thanks for clarifying
